<!-- -->
<launch>



  <!-- carla parameter -->
  <arg name='host' default='localhost'/> <!-- host the carla server is running on -->
  <arg name='port' default='2000'/> <!-- port of the carla server -->
  <arg name="town" default="Town01"/> 
  <!-- <arg name="town" default="/home/adeeb/CARLA_0.9.13/CarlaUE4/Content/Carla/Maps/OpenDrive/no_lane_map.xodr"/> the carla town to load -->
  <arg name='timeout' default='10'/> <!-- the carla timeout for server communication -->
  <arg name='passive' default='False'/> <!-- enable/disable passive mode -->
  <arg name='synchronous_mode_wait_for_vehicle_control_command' default='False'/><!-- should the ros bridge wait for a vehicle control command before proceeding with the next tick -->
  <arg name='synchronous_mode' default='True'/><!-- enable/disable synchronous mode -->
  <arg name='fixed_delta_seconds' default='0.05'/><!-- frequency of the carla ticks -->
  <arg name='register_all_sensors' default='False'/>
  
   
  <arg name="role_name" default="ego_vehicle"/> <!-- role name of the ego vehicle (used for identification in many nodes) -->
  <arg name='spawn_point' default='127.4,-195.4,2,0,0,180'/> 
  <!-- <arg name='spawn_point' default=""/>  -->
  <!-- <arg name='spawn_point' default='71.1,-45.7,35.0,0.0,0.0,180.0'/> -->
  <!-- <arg name="goal_point" default="-64.6,24.4,0.59,0.0,0.0,180"/> goal point for the ego vehicle -->

  <arg name="target_speed" default="8.33"/> <!-- target speed in m/s that the agent tries to follow -->
  <arg name="avoid_risk" default="True"/> <!-- should the agent avoid crashs and respect red traffic lights? -->

  <!-- ===================================================== -->

  <!-- Publish the target speed -->
  <node pkg="rostopic" type="rostopic" name="rostopic" args="pub /carla/$(arg role_name)/target_speed std_msgs/Float64 $(arg target_speed)"/>
  
bddddddd
  <arg name="control_time_step" default="0.05"/>

  <!-- rviz -->
  <node pkg="rviz" type="rviz" name="rviz" args="-d $(find mpc_controller)/config/carla_viz.rviz" output="screen"/>

  <node pkg="global_planner" type="global_path_publisher.py" name="global_path_publisher" output="screen">
    <param name="role_name" value="$(arg role_name)" />
    <param name="host" value="$(arg host)" />
    <param name="port" value="$(arg port)" />
    <param name="timeout" value="$(arg timeout)" />
  </node>

  <node pkg="global_planner" type="frenet_world_converter.py" name="frenet_world_converter" output="screen">
    <param name="role_name" value="$(arg role_name)" />
    <param name="host" value="$(arg host)" />
    <param name="port" value="$(arg port)" />
    <param name="timeout" value="$(arg timeout)" />
  </node>

  <!-- <node pkg="carla_ad_agent" type="ad_agent.py" name="carla_ad_agent_$(arg role_name)" output="screen">
    <param name="role_name" value="$(arg role_name)" />
    <param name="avoid_risk" value="$(arg avoid_risk)" />
  </node> -->

  <node pkg="mpc_controller" type="mpc_controller_mpcc.py" name="mpc_controller_$(arg role_name)" output="screen">
    <param name="role_name" value="$(arg role_name)" />
    <param name="control_time_step" value="$(arg control_time_step)" />
  </node>

  <!-- <include file="$(find carla_ad_demo)/launch/carla_ad_demo.launch">
    <arg name="role_name" value="$(arg role_name)" />
    <arg name="avoid_risk" value="$(arg avoid_risk)" />
  </include> -->

    <!-- The ros bridge -->
  <include file="$(find carla_ros_bridge)/launch/carla_ros_bridge.launch">
    <arg name='host' value='$(arg host)'/>
    <arg name='port' value='$(arg port)'/>
    <arg name='town' value='$(arg town)'/>
    <arg name='timeout' value='$(arg timeout)'/>
    <arg name='passive' value='$(arg passive)'/>
    <arg name='synchronous_mode_wait_for_vehicle_control_command' value='$(arg synchronous_mode_wait_for_vehicle_control_command)'/>
    <arg name='synchronous_mode' value='$(arg synchronous_mode)'/>
    <arg name='fixed_delta_seconds' value='$(arg fixed_delta_seconds)'/>
    <arg name="register_all_sensors" value="$(arg register_all_sensors)"/>
  </include>

  <include file="$(find carla_spawn_objects)/launch/carla_example_ego_vehicle.launch">
    <arg name="objects_definition_file" value='$(find carla_spawn_objects)/config/objects.json'/>
    <arg name='role_name' value='$(arg role_name)'/>
    <arg name="spawn_point_ego_vehicle" value="$(arg spawn_point)"/>
    <arg name="spawn_sensors_only" value="false"/>
  </include>

    <!-- Visualize the ego vehicle (and support overriding the control) -->
  <include file="$(find carla_manual_control)/launch/carla_manual_control.launch">
    <arg name='role_name' value='$(arg role_name)'/>
  </include>

</launch>
